UOP1_Project/Assets/Scripts/StateMachine/Base/StateMachine.cs

```diff
@@ -0,0 +1,208 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace AV.Logic
+{
+    [AddComponentMenu("Logic/State Machine")]
+    public class StateMachine : MonoBehaviour
+    {
+        // TODO: Support multiple states at once.
+        // TODO: StateFlow class for storing and visualizing states relation in a nice editor.
+
+        public StateNode currentState;
+        [HideInInspector]
+        internal bool[] previousDecisions;
+        [HideInInspector]
+        internal new Transform transform;
+
+        // TODO: Runtime data inspector.
+        private Dictionary<Type, Component> cachedComponents;
+        private Dictionary<Type, Component> attachedComponents;
+        private readonly Dictionary<Type, IStateData> attachedData = new Dictionary<Type, IStateData>();
+        
+        private static readonly Dictionary<Type, ScriptableObject> AllSharedData = new Dictionary<Type, ScriptableObject>();
+
+        #if UNITY_EDITOR
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.SubsystemRegistration)]
+        private static void OnDomainReload()
+        {
+            AllSharedData.Clear();
+        }
+        #endif
+        
+        public void Initialize()
+        {
+            transform = base.transform;
+
+            if (!currentState)
+                return;
+
+            InitializeTransitionsData(currentState);
+            CheckStateTriggers(currentState, StateTrigger.OnStart);
+        }
+
+        private void InitializeTransitionsData(StateNode state)
+        {
+            previousDecisions = new bool[state.transitions.Length];
+        }
+        
+        public void Run()
+        {
+            if (!currentState)
+                return;
+
+            currentState.machine = this;
+            currentState.UpdateActions(this);
+            currentState.CheckTransitions(this);
+        }
+
+        public void EnterState(StateNode nextState)
+        {
+            if (!nextState)
+                return;
+            
+            CheckStateTriggers(currentState, StateTrigger.OnExit);
+                
+            currentState = nextState;
+
+            InitializeTransitionsData(nextState);
+            CheckStateTriggers(nextState, StateTrigger.OnEnter);
+        }
+        
+        private void CheckStateTriggers(StateNode state, StateTrigger trigger)
+        {
+            foreach (var action in state.actions)
+            {
+                if (action.trigger != trigger || !action.logic)
+                    continue;
+                action.logic.BeginUpdate(this);
+            }
+        }
+
+        
+        // TODO: Getting data introduces unboxing to generic on every update. How can we avoid this?
```

Comment:
I noticed this in the HungerDecision.OnDecide() implementation, it does makes the 'GetData' call every frame.

Yeah this one's a hard one to figure out. If it's possible to almost add in a 'Init()' method that you'd call once before using the state, that'd be ideal. But since we're using SO's there's no 'Awake()' or 'Start()' we can take advantage of. Quite a pickle!

Reply:
For components, it's easy to add some kind of ComponentPerInstance<T> inside state, which would be generic dictionary that gets data only once. 
But it makes code look harder and more boilerplate:
![image](https://user-images.githubusercontent.com/29812914/95053320-75b0d780-071a-11eb-92b6-b94515397474.png)
It needs to be handles internally, I'll see if I can came up with something better